lib/migrate_user_config.py

```diff
@@ -0,0 +1,166 @@
+    USER_CONFIG_DEFAULTS = {
+        "http_proxy": None,
+        "https_proxy": None,
+        "apt_http_proxy": None,
+        "apt_https_proxy": None,
+        "ua_apt_http_proxy": None,
+        "ua_apt_https_proxy": None,
+        "global_apt_http_proxy": None,
+        "global_apt_https_proxy": None,
+        "update_messaging_timer": 21600,
+        "metering_timer": 14400,
+        "apt_news": True,
+        "apt_news_url": "https://motd.ubuntu.com/aptnews.json",
+    }
+    for field in (
+        "http_proxy",
+        "https_proxy",
+        "apt_http_proxy",
+        "apt_https_proxy",
+        "ua_apt_http_proxy",
+        "ua_apt_https_proxy",
+        "global_apt_http_proxy",
+        "global_apt_https_proxy",
+        "update_messaging_timer",
+        "metering_timer",
+        "apt_news",
+        "apt_news_url",
+    ):
+        old_val = old_user_config.get(field)
+        if old_val is not None and old_val != USER_CONFIG_DEFAULTS.get(field):
+            new_user_config[field] = old_val
+
+    try:
+        with open(
+            "/var/lib/ubuntu-advantage/user-config.json", "w"
+        ) as user_config_file:
+            json.dump(new_user_config, user_config_file)
+    except Exception:
+        if len(new_user_config) > 0:
+            print(
+                messages.USER_CONFIG_MIGRATION_WARNING_NEW_USER_CONFIG_WRITE,
+                file=sys.stderr,
+            )
+            for field in sorted(new_user_config.keys()):
+                print(
+                    "           pro config set {}={}".format(
+                        field, new_user_config[field]
+                    ),
+                    file=sys.stderr,
+                )
+
+
+def create_new_uaclient_conffile(old_uaclient_conf):
+    # Step 3: Create new uaclient.conf
+    # The goal here is to end up with a minimal diff compared to the default
+    # uaclient.conf.
+    # If nothing has changed in any of the uaclient.conf fields, then the
+    # result should exactly match the new default uaclient.conf.
+    new_uaclient_conf = {
+        "contract_url": old_uaclient_conf.get(
+            "contract_url", defaults.BASE_CONTRACT_URL
+        ),
+        "log_level": old_uaclient_conf.get("log_level", "debug"),
+    }
+
+    # these are only present if they were added by the user
+    for field in ("features", "settings_overrides"):
+        if field in old_uaclient_conf:
+            new_uaclient_conf[field] = old_uaclient_conf.get(field)
+
+    # the rest only if they're different from defaults
+    for field in (
+        "data_dir",
+        "log_file",
+        "security_url",
+        "timer_log_file",
+        "daemon_log_file",
+    ):
+        old_val = old_uaclient_conf.get(field)
+        if old_val is not None and old_val != defaults.CONFIG_DEFAULTS.get(
+            field
+        ):
+            new_uaclient_conf[field] = old_val
+
+    try:
+        print(messages.USER_CONFIG_MIGRATION_MIGRATING, file=sys.stderr)
```

Comment:
Should this message be on `stderr` ?

Reply:
I think so, that's where we print some other informational messages in postinst, but I'm not 100% sure

Reply:
I was just wondering because this is not an error message per se, just a status update

Reply:
That's a good point, but is it weird for us to print to stdout in the middle of an upgrade? I don't think we do that anywhere else (but maybe I'm missing something).
@panlinux do you know what is the correct way to display an informational message during an apt upgrade? (OR should we not print anything at all unless something goes wrong?)